File: leetcode/dp/2152_Minimum_Number_of_Lines_to_Cover_Points/codes/dp.c
#include <stdio.h>
#include <limits.h>

int minimumLines(int** points, int pointsSize, int* pointsColSize) {
    int n = pointsSize;
    int maskLine[10][10];
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i == j) {
                maskLine[i][j] = 0;
                continue;
            }
            int mask = 0;
            int x1 = points[i][0], y1 = points[i][1];
            int x2 = points[j][0], y2 = points[j][1];
            for (int k = 0; k < n; k++) {
                int x3 = points[k][0], y3 = points[k][1];
                if ((long long)(x2 - x1) * (y3 - y1) == (long long)(y2 - y1) * (x3 - x1)) {
                    mask |= (1 << k);
                }
            }
            maskLine[i][j] = mask;
        }
    }
    int fullMask = (1 << n) - 1;
    int dp[1 << 10];
    for (int i = 0; i < (1 << n); i++) dp[i] = n;
    dp[0] = 0;
    for (int mask = 0; mask < (1 << n); mask++) {
        if (dp[mask] >= n) continue;
        int first = -1;
        for (int i = 0; i < n; i++) {
            if (!(mask & (1 << i))) {
                first = i;
                break;
            }
        }
        if (first == -1) continue;
        int m1 = mask | (1 << first);
        if (dp[m1] > dp[mask] + 1) dp[m1] = dp[mask] + 1;
        for (int j = 0; j < n; j++) {
            if (j != first && !(mask & (1 << j))) {
                int newMask = mask | maskLine[first][j];
                if (dp[newMask] > dp[mask] + 1) dp[newMask] = dp[mask] + 1;
            }
        }
    }
    return dp[fullMask];
}

File: leetcode/dp/2152_Minimum_Number_of_Lines_to_Cover_Points/codes/dp.py
from typing import List

class Solution:
    def minimumLines(self, points: List[List[int]]) -> int:
        n = len(points)
        maskLine = [[0] * n for _ in range(n)]
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                mask = 0
                x1, y1 = points[i]
                x2, y2 = points[j]
                for k in range(n):
                    x3, y3 = points[k]
                    if (x2 - x1) * (y3 - y1) == (y2 - y1) * (x3 - x1):
                        mask |= (1 << k)
                maskLine[i][j] = mask
        
        fullMask = (1 << n) - 1
        dp = [n] * (1 << n)
        dp[0] = 0
        for mask in range(1 << n):
            if dp[mask] >= n:
                continue
            first = -1
            for i in range(n):
                if not (mask & (1 << i)):
                    first = i
                    break
            if first == -1:
                continue
            dp[mask | (1 << first)] = min(dp[mask | (1 << first)], dp[mask] + 1)
            for j in range(n):
                if j != first and not (mask & (1 << j)):
                    newMask = mask | maskLine[first][j]
                    dp[newMask] = min(dp[newMask], dp[mask] + 1)
        return dp[fullMask]
